ocaml/tests/test_ha_vm_failover.ml

```diff
@@ -569,4 +681,813 @@ module ComputeMaxFailures = Generic.MakeStateful (struct
+
+  let extract_output __context pool =
+    extract_output_for_anti_affinity_plan __context pool
+      Xapi_ha_vm_failover.vm_anti_affinity_spread_evenly_plan
+
+  let tests =
+    `QuickAndAutoDocumented
+      (anti_affinity_plan_test_cases
+      |> List.map (fun (pool, spread_evenly_plan, _no_breach_plan) ->
+             (pool, spread_evenly_plan)
+         )
+      )
+end)
+
+module Slave1EvacuationVMAntiAffinityNoBreachPlan = Generic.MakeStateful (struct
+  module Io = struct
+    type input_t = pool
+
+    type output_t = (string * string) list
+
+    let string_of_input_t = string_of_pool
+
+    let string_of_output_t = Test_printers.(list (pair string string))
+  end
+
+  module State = Test_state.XapiDb
+
+  let load_input __context = setup ~__context
+
+  let extract_output __context pool =
+    extract_output_for_anti_affinity_plan __context pool
+      Xapi_ha_vm_failover.vm_anti_affinity_no_breach_plan
+
+  let tests =
+    `QuickAndAutoDocumented
+      (anti_affinity_plan_test_cases
+      |> List.map (fun (pool, _spread_evenly_plan, no_breach_plan) ->
+             (pool, no_breach_plan)
+         )
+      )
+end)
+
+module Slave1EvacuationPlan = Generic.MakeStateful (struct
+  module Io = struct
+    type input_t = pool
+
+    type output_t = (string * string) list
+
+    let string_of_input_t = string_of_pool
+
+    let string_of_output_t = Test_printers.(list (pair string string))
+  end
+
+  module State = Test_state.XapiDb
+
+  let load_input __context = setup ~__context
+
+  let extract_output __context pool =
+    let total_hosts = Db.Host.get_all ~__context |> List.length in
+    let slv1 =
+      Db.Host.get_all ~__context
+      |> List.find (fun self -> Db.Host.get_name_label ~__context ~self = slave1)
+    in
+    let slave1_vms =
+      Db.Host.get_resident_VMs ~__context ~self:slv1
+      |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+      |> List.filter (fun (_, record) -> not record.API.vM_is_control_domain)
+      |> List.map (fun (self, record) ->
+             (self, Xapi_ha_vm_failover.vm_memory ~__context record)
+         )
+    in
+    let hosts =
+      Db.Host.get_all ~__context
+      |> List.filter (fun h ->
+             Db.Host.get_name_label ~__context ~self:h <> "localhost"
+         )
+      |> List.filter (( <> ) slv1)
+      |> List.map (fun host ->
+             ( host
+             , Memory_check.host_compute_free_memory_with_maximum_compression
+                 ~__context ~host None
+             )
+         )
+    in
+    let vm_can_boot_on_host vm host =
+      let slv1 = pool.slaves |> List.find (fun s -> s.name_label = slave1) in
+      let slave1_vm =
+        slv1.vms
+        |> List.find (fun v ->
+               v.vm_name_label = Db.VM.get_name_label ~__context ~self:vm
+           )
+      in
+      not
+        (List.exists
+           (fun h -> h = Db.Host.get_name_label ~__context ~self:host)
+           slave1_vm.failed_to_boot_on_hosts
+        )
```

Comment:
I compared it to the same function appearing three times, is this function reusable?
```ocaml
    let vm_can_boot_on_host vm host =
      let slv1 = pool.slaves |> List.find (fun s -> s.name_label = slave1) in
      let slave1_vm =
        slv1.vms
        |> List.find (fun v ->
               v.vm_name_label = Db.VM.get_name_label ~__context ~self:vm
           )
      in
      not
        (List.exists
           (fun h -> h = Db.Host.get_name_label ~__context ~self:host)
           slave1_vm.failed_to_boot_on_hosts
        )
```

Reply:
Thanks, move it out and make it used. But I find it was used 2 times.